CONTRIBUTING.md

```diff
@@ -217,27 +217,121 @@ $ export THEIA_PLUGIN_ENDPOINT_DISCOVERY_PORT="2506"
+
+```yaml
+  - mountSources: true
+    memoryLimit: 2Gi
+    type: dockerimage
+    image: 'quay.io/eclipse/che-theia-dev:next'
+    alias: remote-1
+    env:
+      - value: '2504'
+        name: THEIA_PLUGIN_ENDPOINT_DISCOVERY_PORT
+      - value: 'local-dir:///projects/plugins-1/'
+        name: THEIA_PLUGINS
+      - value: /projects/java-11-openjdk
+        name: JAVA_HOME
+```
+
+  Notice the two environment variables we added: `THEIA_PLUGIN_ENDPOINT_DISCOVERY_PORT`and `THEIA_PLUGINS`. The will be used to tell the plugin host process we will run in the container where to announce it's presence and where to pick up the plugins it should run. Note that if you want to make the development process more realistic, you can point the THEIA_PLUGINS variable to some non-shared location, as would be the case in a real Che workspace. Anything in `/tmp` would serve.
+
+We'll also have to add the discovery endpoint setting to the container running the theia backend:
+
+```yaml
+ - mountSources: true
+    endpoints:
+      - name: theia-dev-flow
+        port: 3010
+        attributes:
+          protocol: http
+          public: 'true'
+    memoryLimit: 2Gi
+    type: dockerimage
+    image: 'quay.io/eclipse/che-theia-dev:next'
+    alias: theia-dev
+    env:
+      - value: '2504'
+        name: THEIA_PLUGIN_ENDPOINT_DISCOVERY_PORT
+```
+
+This way, the IDE and the remote plugin host process can find each other.
+
+The container in question must be able to work as a plugin sidecar for any plugin you want to run in it. If you are trying to debug a plugin that already has a sidecar container, it would make sense to base `remote-1`container on the sidecar that runs the original plugin. In that case, you would have to add a copy of node to sidecar container, since they generally don't contain one and we need to run the runtime host process, which is a node application. In this case, however, I've started with a `theia-dev`image and just point it to a JDK I've copied to `/projects/java-11-openjdk`by setting `JAVA_HOME`.
+
+If we want to debug later on, we'll have to add the `node-debug2`Che plugin to our workspace: 
+
+```yaml
+  - type: chePlugin
+    reference: >-
+      https://raw.githubusercontent.com/eclipse/che-plugin-registry/master/v3/plugins/ms-vscode/node-debug2/1.42.3/meta.yaml
+    alias: node-debug
+```
+
+Note that some earlier versions of the plugin did not work in a self-hosting setup, so make sure you are using a recent one. 
+
+## Setting up plugins to run
+
+We can now add plugins to run in the sidecar container by adding them to the folder `/projects/plugins-1`(or whaterver we define as `THEIA_PLUGINS`env variable). If we're just interested in debugging the theia plugins host process, we can download a *.vsix file to the folder: it will be unpacked to a subfolder of `/tmp`. If we need to debug the plugin, we can checkout the source code and build it according the the instructions for the plugin:
```

Comment:
```suggestion
We can now add plugins to run in the sidecar container by adding them to the folder `/projects/plugins-1` (or whaterver we define as `THEIA_PLUGINS` env variable). If we're just interested in debugging the theia plugins host process, we can download a *.vsix file or *.theia file to the folder: it will be unpacked to a subfolder of `/tmp`. If we need to debug the plugin, we can checkout the source code and build it according the the instructions for the plugin:
```

Reply:
Long term, we're trying to de-emphasize the theia API. Plus, the only theia-format plugins I'm aware of are our own, which run in the theia container.